File: runtime_ios_swiftui/ObjectionApp/Object.swift
import Foundation

struct Paragraph: Decodable, Hashable, Equatable {
    let text: String
}

struct Quote: Decodable, Hashable, Equatable {
    let text: String
    let author: String
}

struct ObjectPreview: Decodable, Hashable, Equatable {
    let objectId: String
}

struct CallToAction: Decodable, Hashable, Equatable {
    let title: String
    let icon: String?
    let targetObject: String
}

struct ObjectGroup: Decodable, Hashable, Equatable {
    let title: String
    let description: Bool
    let objectScope: String
}

enum Content: Decodable, Hashable, Equatable {
    case paragraph(Paragraph)
    case quote(Quote)
    case objectPreview(ObjectPreview)
    case callToAction(CallToAction)
    case objectGroup(ObjectGroup)
    
    init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: EnumKeys.self)
        let kind = try container.decode(String.self, forKey: .kind)
        
        switch kind {
        case "paragraph":
            self = .paragraph(try container.decode(Paragraph.self, forKey: .def))
        case "quote":
            self = .quote(try container.decode(Quote.self, forKey: .def))
        case "object_preview":
            self = .objectPreview(try container.decode(ObjectPreview.self, forKey: .def))
        case "call_to_action":
            self = .callToAction(try container.decode(CallToAction.self, forKey: .def))
        case "object_group":
            self = .objectGroup(try container.decode(ObjectGroup.self, forKey: .def))
        default:
            throw DecodingError.dataCorruptedError(forKey: .kind, in: container, debugDescription: "Unknown kind")
        }
    }
}

struct Object: Decodable, Hashable, Equatable {
    let title: String?
    let subtitle: String?
    let description: String?
    let icon: String?
    let content: [Content]
    let actions: [Action]
}

enum ActionKind: String, Hashable, Decodable {
    case danger = "danger"
    case success = "success"
    case normal = "normal"
}

struct Action: Decodable, Hashable, Equatable, Identifiable {
    let id: String
    let kind: ActionKind
    let title: String
    let icon: String?
}

File: runtime_android_compose/app/src/main/java/com/example/objectionapp/Object.kt
package com.example.objectionapp

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator

@Serializable
data class Paragraph(
    @SerialName("text") val text: String
)

@Serializable
data class Quote(
    @SerialName("text") val text: String,
    @SerialName("author") val author: String
)

@Serializable
data class ObjectPreview(
    @SerialName("object_id") val objectId: String
)

@Serializable
data class CallToAction(
    @SerialName("title") val title: String,
    @SerialName("icon") val icon: String? = null,
    @SerialName("target_object") val targetObject: String
)

@Serializable
data class ObjectGroup(
    @SerialName("paragraph") val title: String,
    @SerialName("description") val description: Boolean,
    @SerialName("object_scope") val objectScope: String
)

@OptIn(ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("kind")
sealed class Content {
    @Serializable
    @SerialName("paragraph")
    data class ParagraphContent(val def: Paragraph) : Content()

    @Serializable
    @SerialName("quote")
    data class QuoteContent(val def: Quote) : Content()

    @Serializable
    @SerialName("object_preview")
    data class ObjectPreviewContent(val def: ObjectPreview) : Content()

    @Serializable
    @SerialName("call_to_action")
    data class CallToActionContent(val def: CallToAction) : Content()

    @Serializable
    @SerialName("object_group")
    data class ObjectGroupContent(val def: ObjectGroup) : Content()
}

@Serializable
data class Object(
    @SerialName("title") val title: String? = null,
    @SerialName("subtitle") val subtitle: String? = null,
    @SerialName("description") val description: String? = null,
    @SerialName("icon") val icon: String? = null,
    @SerialName("content") val content: List<Content>,
    @SerialName("actions") val actions: List<Action>
)

@Serializable
enum class ActionKind {
    @SerialName("danger")
    DANGER,

    @SerialName("success")
    SUCCESS,

    @SerialName("normal")
    NORMAL
}

@Serializable
data class Action(
    @SerialName("id") val id: String,
    @SerialName("kind") val kind: ActionKind,
    @SerialName("title") val title: String,
    @SerialName("icon") val icon: String? = null
)
